k8s/policies.go

```diff
@@ -0,0 +1,117 @@
+package k8s
+
+import (
+	"reflect"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	typedCorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const PolicyAnnotationKey = "com.openfaas.policies"
+
+// PolicyClient defines the interface for CRUD operations on policies
+// and applying faaas-netes policies to function Deployments.
+type PolicyClient interface {
+	Get(namespace string, names ...string) ([]Policy, error)
+}
+
+// Policy defined kubernetest specific api extensions that can be predefined and applied
+// to functions by annotating them with `com.openfaas/policy: name1,name2`
+type Policy struct {
+	// If specified, the function's pod tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+// Apply adds or mutates the configuration of the Deployment with the values defined
+// in the Policy. Policies are not merged, if two policies are applied, the last Policy will
+// override preceding Policies with overlapping configurations.
+func (p Policy) Apply(deployment *appsv1.Deployment) *appsv1.Deployment {
+	if len(p.Tolerations) > 0 {
+		deployment.Spec.Template.Spec.Tolerations = p.Tolerations
+	}
+	return deployment
+}
+
+// Remove is the inverse of Apply, removing the mutations that the Policy would have applied
+func (p Policy) Remove(deployment *appsv1.Deployment) *appsv1.Deployment {
+	if reflect.DeepEqual(deployment.Spec.Template.Spec.Tolerations, p.Tolerations) {
+		deployment.Spec.Template.Spec.Tolerations = nil
```

Comment:
I found a bug with this implementation to remove a policy. We can not set Tolerations to `nil` or even to an empty slice, because the field is optional, the empty value seems to be ignored.

Reply:
Odd that you cannot remove / null-out the value? I would look that up on the k/k repo and see if you can find anything there or on Google.